File: src/main/kotlin/solutionsJava/Repetitions.java
import java.util.Scanner;

public class Repetitions {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine().trim();

        System.out.println(longestSubstringWithoutRepeatingChar(str));
    }

    private static int longestSubstringWithoutRepeatingChar(String str) {
        // the string only contains letters ACGT
        // lets map then as 0 1 2 3
        int[] lettersCount = new int[4];

        int left = 0;
        int right = 0;
        int res = 0;
        while (right < str.length()) {
            lettersCount[mIndex(str.charAt(right++))]++;

            if (isValid(lettersCount)) {
                res = Math.max(res, right - left);
            } else {
                while (!isValid(lettersCount)) {
                    lettersCount[mIndex(str.charAt(left++))]--;
                }
            }
        }
        res = Math.max(res, right - left);

        return res;
    }

    private static int mIndex(char c) {
        switch (c) {
            case 'A': return 0;
            case 'C': return 1;
            case 'G': return 2;
            default: return 3;
        }
    }

    private static boolean isValid(int[] lettersCount) {
        int sum = 0;
        for (int count : lettersCount) {
            sum += count;
        }
        for (int count : lettersCount) {
            if (count == sum) return true;
        }
        return false;
    }
}

File: src/main/kotlin/solutions/Repetitions.kt
package uz.safix.solutions

import java.util.Scanner

fun main() {
    val scanner = Scanner(System.`in`)
    val str = scanner.nextLine().trim()

    println(longestSubstringWithoutRepeatingChar(str))
}

private fun longestSubstringWithoutRepeatingChar(str: String): Int {
    // the string only contains letters ACGT
    // lets map then as 0 1 2 3
    val lettersCount = IntArray(4)

    var left = 0
    var right = 0
    var res = 0
    while (right < str.length) {
        lettersCount[str[right++].mIndex]++

        if (lettersCount.isValid()) {
            res = maxOf(res, right - left)
        } else {
            while (lettersCount.isValid().not()) {
                lettersCount[str[left++].mIndex]--
            }
        }
    }
    res = maxOf(res, right - left)

    return res
}

private val Char.mIndex get() = when (this) {
    'A' -> 0
    'C' -> 1
    'G' -> 2
    else -> 3
}

private fun IntArray.isValid(): Boolean {
    val sum = this.sum()
    return this.any { it == sum }
}